n3fit/src/n3fit/hyper_optimization/mongofiletrials.py

```diff
@@ -0,0 +1,290 @@
+        self.database_tar_file = replica_path / f"{self.db_name}.tar.gz"
+        self._parameters = parameters
+        self._rstate = None
+        self._dynamic_trials = []
+
+        super().__init__(self.mongotrials_arg, *args, **kwargs)
+
+    @property
+    def rstate(self):
+        """Returns the rstate attribute; see :class:`n3fit.hyper_optimization.filetrials.FileTrials`."""
+        return self._rstate
+
+    @rstate.setter
+    def rstate(self, random_generator):
+        """Sets the rstate attribute; see :class:`n3fit.hyper_optimization.filetrials.FileTrials`."""
+        self._rstate = random_generator
+
+    def _set_dynamic_trials(self):
+        """Converts self._trials to a dictionary and stores it in self._dynamic_trials."""
+        self._dynamic_trials = [convert_bson_to_dict(item) for item in self._trials]
+
+    def refresh(self):
+        """Fetches data from mongo database and save to a json file."""
+        super().refresh()
+
+        # convert BSON object to a dictionary
+        self._set_dynamic_trials()
+
+        # write json to disk
+        if self._store_trial:
+            log.info("Storing scan in %s", self._json_file)
+            local_trials = []
+            for idx, t in enumerate(self._dynamic_trials):
+                local_trials.append(t)
+                local_trials[idx]["misc"]["space_vals"] = space_eval_trial(self._parameters, t)
+
+            all_to_str = json.dumps(local_trials, default=str)
+            with open(self._json_file, "w") as f:
+                f.write(all_to_str)
+
+    # like in `FileTrials` the two methods below are implemented to avoid writing to the database twice
+    def new_trial_ids(self, n):
+        self._store_trial = False
+        return super().new_trial_ids(n)
+
+    def _insert_trial_docs(self, docs):
+        self._store_trial = True
+        return super()._insert_trial_docs(docs)
+
+    def start_mongo_workers(
+        self, workdir=None, exp_key=None, poll_interval=0.1, use_subprocesses=False
+    ):
+        """Initiates all mongo workers simultaneously."""
+        # get the number of gpu cards, if any
+        gpus_all_physical_list = get_physical_gpus()
+        num_gpus_available = len(gpus_all_physical_list)
+        if not num_gpus_available:
+            log.warning("No GPUs found in the system.")
+
+        # launch mongo workers
+        for i in range(self.num_workers):
+            # construct the command to start a hyperopt-mongo-worker
+            args = [
+                "hyperopt-mongo-worker",
+                "--mongo",
+                f"{self.db_host}:{self.db_port}/{self.db_name}",
+            ]
+            if workdir:
+                args.extend(["--workdir", workdir])
+            if exp_key:
+                args.extend(["--exp-key", exp_key])
+            args.extend(["--poll-interval", str(poll_interval)])
+            if use_subprocesses:
+                args.append("--no-subprocesses")
```

Comment:
I'm confused by this, if you want to use subprocesses, don't use subprocesses? And then it doesn't change the behavior below.
I see this is an option for the mongo worker, which is manually started inside a subprocess. I suppose the reason is to be able to set the GPU it sees through the environment variable? And perhaps these lines are because you don't want to both manually use a subprocess and have mongodb do it as well?

But now `use_subprocesses` defaults to `False` and is never changed, so you are doing both. I have no idea if that's desired or not, but at least this is very confusing.

Reply:
Hi @APJansen. I made a test with `use_subprocesses=True` and it basically means that every worker must wait the other workers to finish the task before going on. You are essentially making the job in serial. That is way this option is turned off by default.

Reply:
Ok, can you then just change `use_subprocesses` to `no_subprocesses`? Otherwise this is confusing.

Reply:
Thanks! Done.